From 287921aad31713e463b91385ab4f6affdda1b4c0 Mon Sep 17 00:00:00 2001
From: DigitalDreamtime <clive.messer@digitaldreamtime.co.uk>
Date: Fri, 6 May 2016 22:22:39 +0100
Subject: [PATCH 05/19] 384k: hifiberry_dacplus: Support 352k8 sample rate

Use CLK44EN for 352k8 sample rate support.

Set (initial) the clk rate to CLK_48EN_RATE in the init method.
Set (reset) the clk rate to CLK_48EN_RATE in the shutdown method.
(Otherwise codec pcm512x_dai_startup_master method can call
 snd_pcm_hw_constraint_ratnums based on CLK_44EN, which will
 result in loss of 384k sample rate, even though we are
 capable of it with CLK_48EN, because 22579200/64=352800.)

Signed-off-by: DigitalDreamtime <clive.messer@digitaldreamtime.co.uk>
---
 sound/soc/bcm/hifiberry_dacplus.c | 25 +++++++++++++++++++++++--
 1 file changed, 23 insertions(+), 2 deletions(-)

diff --git a/sound/soc/bcm/hifiberry_dacplus.c b/sound/soc/bcm/hifiberry_dacplus.c
index bdc35e7..40d7aa4 100644
--- a/sound/soc/bcm/hifiberry_dacplus.c
+++ b/sound/soc/bcm/hifiberry_dacplus.c
@@ -117,6 +117,7 @@ static int snd_rpi_hifiberry_dacplus_clk_for_rate(int sample_rate)
 	case 44100:
 	case 88200:
 	case 176400:
+	case 352800:
 		type = HIFIBERRY_DACPRO_CLK44EN;
 		break;
 	default:
@@ -144,7 +145,7 @@ static void snd_rpi_hifiberry_dacplus_set_sclk(struct snd_soc_codec *codec,
 static int snd_rpi_hifiberry_dacplus_init(struct snd_soc_pcm_runtime *rtd)
 {
 	struct snd_soc_codec *codec = rtd->codec;
-	struct pcm512x_priv *priv;
+	struct pcm512x_priv *priv = snd_soc_codec_get_drvdata(codec);
 
 	if (slave)
 		snd_rpi_hifiberry_is_dacpro = false;
@@ -163,8 +164,16 @@ static int snd_rpi_hifiberry_dacplus_init(struct snd_soc_pcm_runtime *rtd)
 		snd_soc_update_bits(codec, PCM512x_BCLK_LRCLK_CFG, 0x31, 0x11);
 		snd_soc_update_bits(codec, PCM512x_MASTER_MODE, 0x03, 0x03);
 		snd_soc_update_bits(codec, PCM512x_MASTER_CLKDIV_2, 0x7f, 63);
+
+		/*
+		 * Default sclk to CLK_48EN_RATE, otherwise codec
+		 *  pcm512x_dai_startup_master method could call
+		 *  snd_pcm_hw_constraint_ratnums using CLK_44EN/64
+		 *  which will mask 384k sample rate.
+		 */
+		if (!IS_ERR(priv->sclk))
+			clk_set_rate(priv->sclk, CLK_48EN_RATE);
 	} else {
-		priv = snd_soc_codec_get_drvdata(codec);
 		priv->sclk = ERR_PTR(-ENOENT);
 	}
 
@@ -264,6 +273,18 @@ static void snd_rpi_hifiberry_dacplus_shutdown(
 	struct snd_soc_codec *codec = rtd->codec;
 
 	snd_soc_update_bits(codec, PCM512x_GPIO_CONTROL_1, 0x08, 0x00);
+
+	if (snd_rpi_hifiberry_is_dacpro) {
+		struct pcm512x_priv *priv = snd_soc_codec_get_drvdata(codec);
+		/*
+		 * Default sclk to CLK_48EN_RATE, otherwise codec
+		 *  pcm512x_dai_startup_master method could call
+		 *  snd_pcm_hw_constraint_ratnums using CLK_44EN/64
+		 *  which will mask 384k sample rate.
+		 */
+		if (!IS_ERR(priv->sclk))
+			clk_set_rate(priv->sclk, CLK_48EN_RATE);
+	}
 }
 
 /* machine stream operations */
-- 
2.7.4

